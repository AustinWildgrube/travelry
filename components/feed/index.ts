export * from './FeedEnd';
export * from './Post';