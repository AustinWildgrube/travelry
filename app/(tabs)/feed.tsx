import { Fragment } from 'react';
import { Dimensions, FlatList, View, type ListRenderItemInfo } from 'react-native';

import { useInfiniteQuery } from '@tanstack/react-query';
import { Stack } from 'expo-router';
import { Fade, Placeholder, PlaceholderLine, PlaceholderMedia } from 'rn-placeholder';

import { Box } from '&/components/core';
import { FeedEnd, MemoizedPost } from '&/components/feed';
import { Header, LoadingIndicator } from '&/components/shared';
import { getFeedPosts } from '&/queries/post';
import { useUserStore } from '&/stores/user';
import { type PaginatedData, type Post } from '&/types/types';
import { POSTS_PER_PAGE } from '&/utils/constants';
import { makeStyles } from '&/utils/makeStyles';

export default function Feed(): JSX.Element {
  const styles = useStyles();
  const width = Math.round(Dimensions.get('window').width - 32);

  const user = useUserStore(state => state.user);
  if (!user) return <LoadingIndicator />;

  const {
    data: feedPosts,
    isLoading: isLoadingFeedPosts,
    hasNextPage: hasMoreFeedPosts,
    fetchNextPage: fetchNextPageOfPosts,
    refetch: refetchFeedPosts,
    isRefetching: isRefetchingFeedPosts,
  } = useInfiniteQuery({
    queryKey: ['feedPosts'],
    queryFn: ({ pageParam = 1 }) => getFeedPosts(pageParam, user.id),
    getNextPageParam: lastPage => {
      if (lastPage.count && lastPage.count <= lastPage.cursor * POSTS_PER_PAGE - 1) {
        return undefined;
      } else {
        return lastPage.cursor + 1;
      }
    },
  });

  const renderItem = (page: ListRenderItemInfo<PaginatedData<Post>>): JSX.Element => (
    <Fragment>
      {page.item.data.map((post: Post) => (
        <MemoizedPost post={post} key={post.id} />
      ))}
    </Fragment>
  );

  return (
    <View style={styles.feed}>
      <Stack.Screen options={{ headerTitle: () => <Header /> }} />

      {isLoadingFeedPosts &&
        [...Array(3)].map((_, index: number) => (
          <Box key={index} style={styles.placeholderContainer}>
            <Placeholder Animation={Fade}>
              <PlaceholderMedia size={width} />
            </Placeholder>

            <Box style={styles.placeholderDetails}>
              <Placeholder Animation={Fade} Left={PlaceholderMedia}>
                <PlaceholderLine width={80} />
                <PlaceholderLine width={30} />
              </Placeholder>
            </Box>
          </Box>
        ))}

      {!isLoadingFeedPosts && feedPosts && (
        <FlatList
          data={feedPosts.pages}
          renderItem={renderItem}
          refreshing={isRefetchingFeedPosts}
          onRefresh={refetchFeedPosts}
          onEndReached={() => fetchNextPageOfPosts()}
          onEndReachedThreshold={0.5}
          showsVerticalScrollIndicator={false}
          ListFooterComponent={hasMoreFeedPosts ? null : <FeedEnd />}
        />
      )}
    </View>
  );
}

const useStyles = makeStyles(theme => ({
  feed: {
    backgroundColor: theme.colors['backgroundLight0'],
    flex: 1,
    paddingTop: theme.space['2'],
  },
  placeholderContainer: {
    justifyContent: 'center',
    marginHorizontal: theme.space['4'],
    marginTop: theme.space['4'],
  },
  placeholderDetails: {
    marginBottom: theme.space['6'],
    marginTop: theme.space['2'],
  },
}));
