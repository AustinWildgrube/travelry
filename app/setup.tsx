import { Fragment, useState } from 'react';

import { Link } from 'expo-router';
import { SafeAreaView } from 'react-native-safe-area-context';

import { Confetti, SetupProfileDetails, SetupProfileImage } from '&/components/auth';
import { Button, Heading, Text } from '&/components/core';
import { CONFETTI_FALL_SPEED, CONFETTI_ITEM_HEIGHT, CONFETTI_ITEM_WIDTH } from '&/utils/constants';
import { makeStyles } from '&/utils/makeStyles';

export default function Setup(): JSX.Element {
  const styles = useStyles();
  const [step, setStep] = useState(0);

  return (
    <SafeAreaView style={styles.setup}>
      <Heading>Setup your profile</Heading>

      {step === 0 ? (
        <Fragment>
          <Text style={styles.setupSlogan}>Upload an image - it's worth a thousand words</Text>
          <SetupProfileImage setStep={setStep} />

          <Button onPress={() => setStep(1)} variant="link" style={styles.skipLink}>
            <Button.Text>Skip</Button.Text>
          </Button>

          <Confetti
            fallSpeed={CONFETTI_FALL_SPEED}
            flipSpeed={3}
            horizSpeed={50}
            itemColors={['#84a98c', '#ffb703', '#023047']}
            itemDimensions={{ height: CONFETTI_ITEM_HEIGHT, width: CONFETTI_ITEM_WIDTH }}
            itemTintStrength={0.8}
            numItems={100}
          />
        </Fragment>
      ) : (
        <Fragment>
          <Text style={styles.setupSlogan}>Tell the world a little about yourself</Text>
          <SetupProfileDetails />

          <Link href="/feed" asChild>
            <Button variant="link" style={styles.skipLink}>
              <Button.Text>Skip</Button.Text>
            </Button>
          </Link>
        </Fragment>
      )}
    </SafeAreaView>
  );
}

const useStyles = makeStyles(theme => ({
  setup: {
    flex: 1,
    padding: theme.space['4'],
  },
  setupSlogan: {
    marginTop: theme.space['1'],
    marginBottom: theme.space['6'],
  },
  skipLink: {
    marginTop: theme.space['3'],
    zIndex: -1,
  },
}));
