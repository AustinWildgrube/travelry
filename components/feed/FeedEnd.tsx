import { Alert } from 'react-native';

import { Button, Text, VStack } from '&/components/core';
import { makeStyles } from '&/utils/makeStyles';

export function FeedEnd(): JSX.Element {
  const styles = useStyle();

  const triggerAlert = (): void => {
    Alert.alert(
      'Why am I seeing this?',
      "As our community is just getting started, we've reached the end of our current content. But don't worry, there's still plenty of " +
        "room for you to help us grow! We encourage you to start sharing your own travel experiences and inspire others to do the same. Let's " +
        'build our community together!',
    );
  };

  return (
    <VStack style={styles.feedEnd}>
      <Text>You've reached the end.</Text>

      <Button onPress={triggerAlert} variant="link">
        <Button.Text>Wonder why you're seeing this?</Button.Text>
      </Button>
    </VStack>
  );
}

const useStyle = makeStyles(theme => ({
  feedEnd: {
    alignItems: 'center',
    marginVertical: theme.space['4'],
  },
}));
