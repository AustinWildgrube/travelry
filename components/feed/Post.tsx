import { memo, useState } from 'react';
import { ImageBackground } from 'react-native';

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { impactAsync, ImpactFeedbackStyle } from 'expo-haptics';
import { LinearGradient } from 'expo-linear-gradient';
import { Link, useRouter } from 'expo-router';

import { Avatar, Heading, HStack, Pressable, Text, VStack } from '&/components/core';
import { LoadingIndicator } from '&/components/shared/LoadingIndicator';
import { useAvatarQuery } from '&/hooks/useAvatarQuery';
import { useDoubleTap } from '&/hooks/useDoubleTap';
import { likePost, unlikePost } from '&/queries/like';
import { usePostStore } from '&/stores/post';
import { useUserStore } from '&/stores/user';
import { type InfiniteQuery, type PaginatedData, type Post } from '&/types/types';
import { getPublicImageUrl } from '&/utils/getPublicImageUrl';
import { makeStyles } from '&/utils/makeStyles';

type PostProps = {
  post: Post;
};

function Post({ post }: PostProps): JSX.Element {
  const styles = useStyles();
  const router = useRouter();
  const queryClient = useQueryClient();
  const setViewedPost = usePostStore(state => state.setViewedPost);

  const [isPostLiked, setIsPostLiked] = useState(Object.keys(post.post_like).length > 0);

  const user = useUserStore(state => state.user);
  if (!user) return <LoadingIndicator />;

  const doubleTapBind = useDoubleTap(() => mutatePostLike(), {
    onSingleTap: () => {
      setViewedPost(post);
      router.push({ pathname: `/post/${post.id}` });
    },
  });

  const { data: avatar } = useAvatarQuery(post.account.id, post.account.avatar_url);

  const { data: postMedia } = useQuery({
    queryKey: ['postMedia', post.id],
    queryFn: () => getPublicImageUrl('posts', post.post_media[0].file_url),
  });

  const { mutate: mutatePostLike } = useMutation({
    mutationFn: () => (isPostLiked ? unlikePost(user.id, post.id) : likePost(user.id, post.id)),
    onMutate: () => impactAsync(ImpactFeedbackStyle.Medium),
    onSuccess: async () => {
      await queryClient.setQueryData(['feedPosts'], (oldData: InfiniteQuery<PaginatedData<Post>> | undefined): InfiniteQuery<PaginatedData<Post>> => {
        if (!oldData) return { pageParams: [], pages: [{ count: 0, cursor: 0, data: [] }] };

        for (const page of oldData.pages) {
          for (const oldPost of page.data) {
            if (oldPost.id === post.id) {
              oldPost.post_stat.likes_count = isPostLiked ? oldPost.post_stat.likes_count - 1 : oldPost.post_stat.likes_count + 1;
              break;
            }
          }
        }

        return oldData;
      });

      setIsPostLiked(!isPostLiked);
    },
  });

  return (
    <Pressable key={post.created_at} style={styles.post} {...doubleTapBind}>
      <ImageBackground source={{ uri: postMedia }} style={styles.postMedia}>
        <LinearGradient colors={['rgba(0, 0, 0, 0.8)', 'transparent']} start={{ x: 0, y: 0 }} end={{ x: 0, y: 1 }} style={styles.postMediaGradient} />

        <Link href={`/(tabs)/${post.account.id}`} style={styles.postInfo} asChild>
          <HStack as={Pressable} space="sm" style={styles.postInfoWrapper}>
            <Avatar style={styles.postAvatar}>
              <Avatar.FallbackText>
                {post.account.first_name} {post.account.last_name}
              </Avatar.FallbackText>

              <Avatar.Image source={{ uri: avatar }} placeholder={post.account.avatar_placeholder} />
            </Avatar>

            <VStack>
              <Heading size="xs" style={styles.postInfoText}>
                {post.account.first_name} {post.account.last_name}
              </Heading>

              <Text size="xs" style={styles.postInfoText}>
                {post.location}
              </Text>
            </VStack>
          </HStack>
        </Link>
      </ImageBackground>
    </Pressable>
  );
}

const useStyles = makeStyles(theme => ({
  post: {
    margin: theme.space['0.5'],
    marginHorizontal: theme.space['1'],
  },
  postMedia: {
    alignItems: 'flex-start',
    borderRadius: theme.radii['md'],
    height: theme.space['96'],
    flexDirection: 'row',
    justifyContent: 'space-between',
    padding: theme.space['2'],
    overflow: 'hidden',
  },
  postMediaGradient: {
    height: '25%',
    left: 0,
    position: 'absolute',
    right: 0,
    top: 0,
  },
  postInfo: {
    padding: theme.space['2'],
  },
  postInfoWrapper: {
    alignItems: 'center',
    marginTop: theme.space['2'],
  },
  postAvatar: {
    height: theme.space['10'],
    width: theme.space['10'],
  },
  postInfoText: {
    color: theme.colors['white'],
  },
}));

export const MemoizedPost = memo(Post);
