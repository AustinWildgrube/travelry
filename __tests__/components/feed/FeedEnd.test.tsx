import { Alert } from 'react-native';

import { fireEvent, screen } from '@testing-library/react-native';

import { FeedEnd } from '&/components/feed/FeedEnd';

import { render } from '&/jest.setup';

jest.spyOn(Alert, 'alert');

describe('FeedEnd', () => {
  beforeEach(() => {
    render(<FeedEnd />);
  });

  it('should match the snapshot', () => {
    screen.toJSON();
    expect(screen).toMatchSnapshot();
  });

  it('should render the button correctly', () => {
    const endOfContentText = screen.getByText(/You've reached the end./i);
    const button = screen.getByRole('button', { name: /Wonder why you're seeing this?/i });

    expect(endOfContentText).toBeTruthy();
    expect(button).toBeTruthy();
  });

  it('shows alert when button is pressed', () => {
    const button = screen.getByRole('button', { name: /Wonder why you're seeing this?/i });
    fireEvent.press(button);

    expect(Alert.alert).toHaveBeenCalledWith(
      'Why am I seeing this?',
      "As our community is just getting started, we've reached the end of our current content. But don't worry, there's still plenty of " +
        "room for you to help us grow! We encourage you to start sharing your own travel experiences and inspire others to do the same. Let's " +
        'build our community together!',
    );
  });
});
