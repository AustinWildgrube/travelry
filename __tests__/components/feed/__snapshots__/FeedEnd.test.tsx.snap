// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedEnd should match the snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#ffffff",
        "flex": 1,
        "zIndex": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "paddingVertical": 24,
        }
      }
    >
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#242424",
            "fontFamily": "Inter_400Regular",
            "fontSize": 14,
            "fontWeight": "400",
            "lineHeight": 20,
          }
        }
      >
        You've reached the end.
      </Text>
      <View
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 2,
            },
            undefined,
            false,
          ]
        }
        testID="RNE_BUTTON_WRAPPER"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": false,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderColor": "#84a98c",
                "borderRadius": 2,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
                "paddingHorizontal": 10,
                "paddingTop": 4,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#84a98c",
                  "fontSize": 14,
                  "paddingTop": 0,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Wonder why you're seeing this?
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
